### PR TITLE
Feature #90 custom field in resource groups

### PR DESCRIPTION
**Pull request by EliasBoulharts**
Issue: #90

### CODE REVIEW

File: resource_tracker/forms/resource_group_attribute_definition_form.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+                                         widget=forms.Select(attrs={'class': 'form-control'}))
+
+    def clean(self):
+        name = self.cleaned_data['name']
+        produce_for = self.cleaned_data['produce_for']
+        consume_from = self.cleaned_data['consume_from']
+        if not self.instance.id:
+            try:
+                self.resource_group.add_attribute_definition(name, produce_for, consume_from)
+            except ExceptionResourceTracker.AttributeAlreadyExist as e:
```
Comment (Sispheor) [inline resource_tracker/forms/resource_group_attribute_definition_form.py:33]: missing space avec ","

File: templates/resource_tracking/resource_group/resources/resource-list.html [hunk 2]
```diff
@@ -29,6 +29,9 @@
                                 {% for attribute_name in list_attribute_name %}
                                     <th>{{ attribute_name }}</th>
                                 {%  endfor %}
+                                {% for attribute_name in list_text_attribute_name %}
+                                    <th>{{ attribute_name }} (text)</th>
```
Comment (Sispheor) [inline templates/resource_tracking/resource_group/resources/resource-list.html:33]: remove "(text)".
To verbose

File: resource_tracker/forms/resource_form.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+                try:
+                    initial_value = ResourceTextAttribute.objects.get(resource=self.instance,
+                                                                      text_attribute_type=text_attribute)
+                except ResourceTextAttribute.DoesNotExist:
+                    pass
+            new_field = forms.CharField(label=text_attribute.name,
+                                        max_length=500,
+                                        required=False,
+                                        initial=initial_value,
+                                        help_text="Text field",
```
Comment (Sispheor) [inline resource_tracker/forms/resource_form.py:58]: The help text here should be set by the user. (non mandatory field on attribute creation)

File: resource_tracker/models/resource_group.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
+from django.db import models
+from taggit.managers import TaggableManager
+
+from resource_tracker.models import ResourceGroupAttributeDefinition, ExceptionResourceTracker, ResourceTextAttribute, \
+    ResourceAttribute, ResourceGroupTextAttributeDefinition
+
+
+class ResourceGroup(models.Model):
```
Comment (Sispheor) [inline resource_tracker/models/resource_group.py:8]: seems that all method of this models are not tested

File: resource_tracker/models/resource_group_attribute_definition.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    produce_for = models.ForeignKey('ResourcePoolAttributeDefinition',
+                                    on_delete=models.PROTECT,
+                                    related_name='producers',
+                                    related_query_name='producer',
+                                    null=True)
+
+    def __str__(self):
+        return f"{self.resource_group_definition} - {self.name}"
+
+    def get_total_resource(self):
```
Comment (Sispheor) [inline resource_tracker/models/resource_group_attribute_definition.py:28]: not tested method

File: resource_tracker/models/resource_pool_attribute_definition.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    def get_total_consumed_by(self, consumer):
+        total_consumed = 0
+        for resource in consumer.resource_group_definition.resources.all():
+            try:
+                total_consumed += resource.attributes.get(attribute_type=consumer).value
+            except ResourceAttribute.DoesNotExist:
+                pass
+        return total_consumed * self.over_commitment_consumers
+
+    def remove_all_producer(self):
```
Comment (Sispheor) [inline resource_tracker/models/resource_pool_attribute_definition.py:80]: remove_all_producer and remove_all_consumer are not tested

**Review by Sispheor — CHANGES_REQUESTED** [review]
Please add as well a buton to go to resource list from the resource group edit page

File: resource_tracker/forms/resource_group_attribute_definition_form.py [hunk 7]
```diff
@@ -1,16 +1,16 @@
...
 
 
 class ResourceGroupAttributeDefinitionForm(ModelForm):
     class Meta:
         model = ResourceGroupAttributeDefinition
-        fields = ["name", "consume_from", "produce_for"]
+        fields = ["name", "help_text", "consume_from", "produce_for"]
 
     name = forms.CharField(label="Name",
-                           required=True,
```
Comment (a-belhadj) [inline resource_tracker/forms/resource_group_attribute_definition_form.py]: Default value, useless

File: resource_tracker/forms/resource_form.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
+from django.forms import ModelForm
+from taggit.forms import *
+from resource_tracker.models import ResourceGroup, ResourceGroupAttributeDefinition, Resource, ResourceAttribute, \
+    ResourceTextAttribute, ResourceGroupTextAttributeDefinition
+from service_catalog.models import Instance
+
+
+class ResourceForm(ModelForm):
```
Comment (Sispheor) [inline resource_tracker/forms/resource_form.py:8]: Dedicated test to be done

File: resource_tracker/forms/resource_group_attribute_definition_form.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+class ResourceGroupAttributeDefinitionForm(ModelForm):
+    class Meta:
+        model = ResourceGroupAttributeDefinition
+        fields = ["name", "help_text", "consume_from", "produce_for"]
+
+    name = forms.CharField(label="Name",
+                           widget=forms.TextInput(attrs={'class': 'form-control'}))
+
+    consume_from = forms.ModelChoiceField(queryset=ResourcePoolAttributeDefinition.objects.all(),
+                                          to_field_name="id",
```
Comment (Sispheor) [inline resource_tracker/forms/resource_group_attribute_definition_form.py:17]: this is the default behavior of Django

File: resource_tracker/forms/resource_group_attribute_definition_form.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+                                         required=False,
+                                         widget=forms.Select(attrs={'class': 'form-control'}))
+
+    help_text = forms.CharField(label="Help text",
+                                required=False,
+                                max_length=ResourceGroupAttributeDefinition._meta.get_field('help_text').max_length,
+                                widget=forms.TextInput(attrs={'class': 'form-control'}))
+
+    def clean(self):
+        name = self.cleaned_data['name']
```
Comment (Sispheor) [inline resource_tracker/forms/resource_group_attribute_definition_form.py:32]: missing super(ModelForm)

File: resource_tracker/models/resource_group_attribute_definition.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
+from django.db import models
+
+from resource_tracker.models import ResourceAttribute
+
+
+class ResourceGroupAttributeDefinition(models.Model):
+    name = models.CharField(max_length=100,
+                            blank=False)
+    resource_group_definition = models.ForeignKey('ResourceGroup',
```
Comment (Sispheor) [inline resource_tracker/models/resource_group_attribute_definition.py:9]: spaces

File: resource_tracker/tests/test_resource_group_text_attribute_views.py [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+        data = {
+            "name": "new resource"
+        }
+        response = self.client.post(url, data=data)
+        self.assertEquals(302, response.status_code)
+
+        url = reverse('resource_tracker:resource_group_resource_list', kwargs=args)
+        response = self.client.get(url, data=data)
+        self.assertEquals(200, response.status_code)
+
```
Comment (Sispheor) [inline resource_tracker/tests/test_resource_group_text_attribute_views.py:168]: add test to check that the new object has been created in the DB

**Review by Sispheor — APPROVED** [review]

**Pull request merged**